### PR TITLE
Haskell GHC JSON format support (fixes #182)

### LINKED ISSUE #182: Format support: Haskell GHC JSON Profile
See [the Haskell profiling documentation](https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/profiling.html) for a description of the `-pj` JSON format. It won't support the time sort view since the profile only includes a summary and not the individual samples.

I currently plan on implementing this myself and submitting a PR.

### PR DESCRIPTION
**Pull request by trishume**
Fixes #182 by adding support for importing the JSON profiling format created by GHC's built in profiling support when the executable is passed the `-pj` option. Produces a profile group containing both a time and allocation profile.

Unfortunately, GHC doesn't provide the raw sample information to get the time view to be useful, so only left heavy and sandwich are useful.

Includes a test profile, and I've also tested it on a more real large 2MB profile file in the UI and it works great.

I also modified the Readme to link to a wiki page I'm unable to create, but that should have something like this content copy-pasted into it:

# Importing from Haskell

GHC provides built in profiling support that can export a JSON file.
In order to do this you need to compile your executable with profiling
support and then pass the `-pj` RTS flag to the executable.

This will produce a `my-binary.prof` file in the current directory which
you can import into speedscope.

## Using GHC

See the [GHC manual page on profiling](https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/profiling.html)
for more extensive information on the command line flags available.

```
$ ghc -prof -fprof-auto -rtsopts Main.hs
$ ./Main +RTS -pj -RTS
```

### With executables

```
$ stack build --profile
$ stack exec -- my-executable +RTS -pj -RTS
```

### With tests

```
stack test --profile --test-arguments "+RTS -pj -RTS"
```

### CODE REVIEW

File: src/import/haskell.ts [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+      frameInfo.file = centre.src_loc
+    }
+
+    idToFrameInfo.set(centre.id, frameInfo)
+  }
+
+  const timeProfile = new CallTreeProfileBuilder(haskellProfile.total_ticks)
+  addToProfile(haskellProfile.profile, 0, timeProfile, idToFrameInfo, tree => tree.ticks)
+  timeProfile.setValueFormatter(new TimeFormatter('milliseconds'))
+  console.log(timeProfile.getName())
```
Comment (jlfwong) [inline src/import/haskell.ts]: Was this left here intentionally?

```suggestion
```

File: src/import/haskell.ts [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+export function importFromHaskell(haskellProfile: HaskellProfile): ProfileGroup {
+  const idToFrameInfo = new Map<number, FrameInfo>()
+  for (let centre of haskellProfile.cost_centres) {
+    const frameInfo: FrameInfo = {
+      key: centre.id,
+      name: `${centre.module}.${centre.label}`,
+    }
+
+    // ignore things like <entire-module> and <no location info>
```
Comment (jlfwong) [inline src/import/haskell.ts]: A small style nit: I try to have comments capitalized

```suggestion
    // Ignore things like <entire-module> and <no location info>
```

File: src/import/haskell.ts [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    if (!centre.src_loc.startsWith('<')) {
+      // this also contains line and column information, but sometimes it contains ranges,
```
Comment (jlfwong) [inline src/import/haskell.ts]:
```suggestion
      // This also contains line and column information, but sometimes it contains ranges,
```

**Comment by trishume** [comment]
Thanks, suggestions applied, such a great new feature. 

Although I wish there was a way to apply multiple suggestions in one commit 😐. I wonder if until they add that ability there will be a surge in maintainers using the squash before merging strategy...

**Comment by jlfwong** [comment]
Thanks @trishume! This is now merged, published to npm, and deployed to https://www.speedscope.app/. I also added the documentation you wrote to the wiki: https://github.com/jlfwong/speedscope/wiki/Importing-from-Haskell

**Pull request merged**